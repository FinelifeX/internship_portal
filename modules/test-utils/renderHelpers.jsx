/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import TR from 'react-test-renderer';
import { render as RTLRender } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

/**
 * @typedef {Object.<string, any>} Props
 */

/**
 * @typedef {React.ReactNode} Children
 */

/**
 * @typedef {{props?: Props, children?: Children}} RenderOptions
 */

/**
 *
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @returns {JSX.Element}
 */
export const render = (Component, options = {}) => {
  const { props, children } = options;
  return <Component {...props}>{children}</Component>;
};

/**
 * Render React component inside MemoryRouter
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @param {import('react-router-dom').MemoryRouterProps} routerOptions
 * @returns {JSX.Element}
 */
export const renderWithRouter = (
  Component,
  options = {},
  routerOptions = {},
) => {
  const { props, children } = options;

  return (
    <MemoryRouter {...routerOptions}>
      <Component {...props}>{children}</Component>
    </MemoryRouter>
  );
};

/**
 * Shallow render component using "react-test-renderer"
 * @type {import('react-test-renderer').create}
 */
export const shallow = TR.create;

/**
 * Render component using "render()" from "@testing-library/react"
 * @type {import('@testing-library/react').render}
 */
export const mount = RTLRender;
