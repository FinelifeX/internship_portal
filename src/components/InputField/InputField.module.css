.input {
  padding: 8px 12px;
  font-size: 16px;
  line-height: 24px;
  width: 100%;
  border: 1px solid #d9d9d9;
  border-radius: 2px;
}

.input::placeholder {
  color: rgba(0, 0, 0, 0.25);
}
