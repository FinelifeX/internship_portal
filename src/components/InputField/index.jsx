import React from 'react';
import PT from 'prop-types';
import classNames from 'classnames';
import styles from './InputField.module.css';

const InputField = ({
  className,
  id,
  type,
  value,
  placeholder,
  onChange,
  onFocus,
  onBlur,
}) => (
  <input
    className={classNames(styles.input, className)}
    id={id}
    type={type}
    value={value}
    placeholder={placeholder}
    onChange={onChange}
    onFocus={onFocus}
    onBlur={onBlur}
  />
);

InputField.defaultProps = {
  className: '',
  id: undefined,
  type: 'text',
  value: undefined,
  placeholder: '',
  onChange: () => {},
  onFocus: () => {},
  onBlur: () => {},
};

InputField.propTypes = {
  className: PT.string,
  id: PT.oneOfType([PT.string, PT.number]),
  type: PT.oneOf(['text', 'password', 'email', 'number']),
  value: PT.oneOfType([PT.number, PT.string]),
  placeholder: PT.string,
  onChange: PT.func,
  onFocus: PT.func,
  onBlur: PT.func,
};

export default InputField;
