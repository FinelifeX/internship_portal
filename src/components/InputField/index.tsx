import React, { ChangeEventHandler, FC, FocusEventHandler } from 'react';
import classNames from 'classnames';
import styles from './InputField.module.css';

export type InputFieldProps = {
  className?: string;
  id?: string;
  type?: 'text' | 'password' | 'email' | 'number';
  value?: string | number;
  placeholder?: string;
  onChange: ChangeEventHandler<HTMLInputElement>;
  onFocus: FocusEventHandler<HTMLInputElement>;
  onBlur: FocusEventHandler<HTMLInputElement>;
};
const InputField: FC<InputFieldProps> = ({
  className,
  id,
  type,
  value,
  placeholder,
  onChange,
  onFocus,
  onBlur,
}) => (
  <input
    data-testid="input"
    className={classNames(styles.input, className)}
    id={id}
    type={type}
    value={value}
    placeholder={placeholder}
    onChange={onChange}
    onFocus={onFocus}
    onBlur={onBlur}
  />
);

export default InputField;
