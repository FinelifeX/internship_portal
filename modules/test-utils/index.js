import * as RenderHelpers from './renderHelpers';

export { RenderHelpers };
