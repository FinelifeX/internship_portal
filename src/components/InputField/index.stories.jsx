import React from 'react';
import InputField from './index';

export default {
  component: InputField,
  title: 'Components/InputField',
};

const Template = (args) => <InputField {...args} />;

export const Default = Template.bind({});
