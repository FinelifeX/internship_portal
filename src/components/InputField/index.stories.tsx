import React from 'react';
import InputField, { InputFieldProps } from './index';

export default {
  component: InputField,
  title: 'Components/InputField',
};

const Template = (props: InputFieldProps) => <InputField {...props} />;

export const Default = Template.bind({});
