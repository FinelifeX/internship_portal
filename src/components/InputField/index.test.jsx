import { fireEvent } from '@testing-library/react';
import { shallow, mount, render } from 'test-utils/renderHelpers';
import InputField from './index';

describe('InputField component', () => {
  const TEST_ID = 'input';
  const defaultProps = {
    onChange: jest.fn(),
    onFocus: jest.fn(),
    onBlur: jest.fn(),
  };

  it('should render correctly', () => {
    const tree = shallow(render(InputField, { defaultProps })).toJSON();

    expect(tree).toMatchSnapshot();
  });

  const renderComponent = (props) =>
    mount(render(InputField, { defaultProps, props }));

  it('should have value that was passed via props', () => {
    const props = { value: '123' };
    const { getByTestId } = renderComponent(props);

    expect(getByTestId(TEST_ID)).toHaveValue(props.value);
  });

  it('should call onFocus when focused', () => {
    const { getByTestId } = renderComponent();

    fireEvent.focus(getByTestId(TEST_ID));

    expect(defaultProps.onFocus).toBeCalledTimes(1);
  });

  it('should call onBlur when unfocused', () => {
    const { getByTestId } = renderComponent();

    fireEvent.blur(getByTestId(TEST_ID));

    expect(defaultProps.onBlur).toBeCalledTimes(1);
  });

  it('should call onChange with provided value when change occurs', () => {
    const { getByTestId } = renderComponent();

    fireEvent.change(getByTestId(TEST_ID), { target: { value: '123' } });

    expect(defaultProps.onChange).toBeCalledTimes(1);
  });
});
