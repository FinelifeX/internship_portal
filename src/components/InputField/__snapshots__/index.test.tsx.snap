// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField component should render correctly 1`] = `
<input
  className="input"
  data-testid="input"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  onFocus={[MockFunction]}
/>
`;
